SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/VisibilityTest/VisibilityTest.cpp

```diff
@@ -0,0 +1,341 @@
+ *    - The controllers for each client  possess the spawned test character pawns
+ *  - Test:
+ *    - Server moves one Client to a remote position where Client can not see the HiddenActor(Cube) .
+ *	  - Check if the Clients can see the HiddenActor(Cube).
+ *	  - Server sets the HiddenActor(Cube) as bHidden.
+ *	  - Server moves the remote Client back close to the HiddenActor(Cube).
+ *	  - Check if the Clients can see the HiddenActor(Cube).
+ *	  - Server sets the HiddenActor(Cube) as Visible (bHidden = false).
+ *	  - Check if the Clients can see the HiddenActor(Cube).
+ *  - Cleanup:
+ *    - The clients repossess their default pawns
+ *    - The test pawns are destroyed
+ */
+
+AVisibilityTest::AVisibilityTest()
+	: Super()
+{
+	Author = "Evi";
+	Description = TEXT("Test Actor Visibility");
+}
+
+void AVisibilityTest::GetLifetimeReplicatedProps(TArray<FLifetimeProperty>& OutLifetimeProps) const
+{
+	Super::GetLifetimeReplicatedProps(OutLifetimeProps);
+
+	DOREPLIFETIME(AVisibilityTest, Controllers);
+	DOREPLIFETIME(AVisibilityTest, TestPawns);
+}
+
+void AVisibilityTest::BeginPlay()
+{
+	Super::BeginPlay();
+
+	CharacterRemoteLocation = FVector(20000.0f, 20000.0f, 50.0f);
+	Character1StartingLocation = FVector(0.0f, 120.0f, 50.0f);
+
+	{	// Step 1 - Set up the world Spawn PlayerCharacter and make sure the cube is in the world.
+		AddStep(TEXT("ServerSetup"), FWorkerDefinition::Server(1), nullptr, [this](ASpatialFunctionalTest* NetTest)
+		{
+			int Counter = 0;
+			int ExpectedReplicatedActors = 1;
+			FVector StartingLocation = FVector::ZeroVector;
+			float YToSpawnAt = 120.0f;
+			float YSpawnIncrement = 120.0f;
+
+			TestPawns.Empty();
+			Controllers.Empty();
+
+			for (TActorIterator<AReplicatedVisibilityTestActor> Iter(GetWorld()); Iter; ++Iter)
+			{
+				Counter++;
+			}
+			AssertEqual_Int(Counter, ExpectedReplicatedActors, TEXT("Number of TestHiddenActors in the server world"), NetTest);
+
+			for (ASpatialFunctionalTestFlowController* FlowController : GetFlowControllers())
+			{
+				checkf(FlowController, TEXT("Can't be running test without valid FlowControl."));
+
+				if (FlowController->WorkerDefinition.Type == ESpatialFunctionalTestWorkerType::Server)
+				{
+					continue;
+				}
+				
+				StartingLocation = FVector(0.0f, YToSpawnAt, 50.0f);
+				ATestMovementCharacter* TestCharacter = GetWorld()->SpawnActor<ATestMovementCharacter>(StartingLocation, FRotator::ZeroRotator, FActorSpawnParameters());
+				YToSpawnAt += YSpawnIncrement;
+
+				APlayerController* PlayerController = Cast<APlayerController>(FlowController->GetOwner());
+				Controllers.Add(PlayerController);
+
+				TestPawns.Add(TestCharacter);
+
+				RegisterAutoDestroyActor(TestCharacter);
+
+				// Save old one to put it back in the final step
+				OriginalPawns.Add(TPair<AController*, APawn*>(PlayerController, PlayerController->GetPawn()));
+
+				PlayerController->Possess(TestCharacter);
+
+			}
```

Comment:
Correct me if I am wrong but it seems that only Client 1's Character is moved, right? If that's the case, then I think you can replace this bit of code by using the same structure as in the image below. (This is something I have recently realised and also changed in my tests 😄). And if that is the case, then you can also remove the 2 floats YToSpawnAt and the YSpawnIncrement. Again, this is my personal preference, but it seems to make the code a bit easier to understand in my view.

![image](https://user-images.githubusercontent.com/66263711/89892790-209b8e00-dbcf-11ea-8601-53ddb5fd2144.png)